{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'c:\\\\Users\\\\manob\\\\Documents\\\\Kidney_Disease_Detection\\\\Kidney_Disease_Detection\\\\research'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"../\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'c:\\\\Users\\\\manob\\\\Documents\\\\Kidney_Disease_Detection\\\\Kidney_Disease_Detection'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dataclasses import dataclass\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "@dataclass(frozen=True)\n",
    "class TrainingConfig:\n",
    "    root_dir: Path\n",
    "    trained_model_path: Path\n",
    "    updated_base_model_path: Path\n",
    "    training_data: Path\n",
    "    params_epochs: int\n",
    "    params_batch_size: int\n",
    "    params_is_augmentation: bool\n",
    "    params_image_size: list\n",
    "    params_learning_rate: float "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\manob\\Documents\\Kidney_Disease_Detection\\Kidney_Disease_Detection\\.conda\\Lib\\site-packages\\numpy\\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:\n",
      "c:\\Users\\manob\\Documents\\Kidney_Disease_Detection\\Kidney_Disease_Detection\\.conda\\Lib\\site-packages\\numpy\\.libs\\libopenblas.FB5AE2TYXYH2IJRDKGDGQ3XBKLKTF43H.gfortran-win_amd64.dll\n",
      "c:\\Users\\manob\\Documents\\Kidney_Disease_Detection\\Kidney_Disease_Detection\\.conda\\Lib\\site-packages\\numpy\\.libs\\libopenblas64__v0.3.21-gcc_10_3_0.dll\n",
      "  warnings.warn(\"loaded more than 1 DLL from .libs:\"\n"
     ]
    }
   ],
   "source": [
    "from cnnClassifier.constants import *\n",
    "from cnnClassifier.utils.common import read_yaml, create_directories\n",
    "import tensorflow as tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ConfigurationManager:\n",
    "    def __init__(\n",
    "        self,\n",
    "        config_filepath = CONFIG_FILE_PATH,\n",
    "        params_filepath = PARAMS_FILE_PATH):\n",
    "\n",
    "        self.config = read_yaml(config_filepath)\n",
    "        self.params = read_yaml(params_filepath)\n",
    "\n",
    "        create_directories([self.config.artifacts_root])\n",
    "\n",
    "\n",
    "    def get_training_config(self) -> TrainingConfig:\n",
    "        training = self.config.training\n",
    "\n",
    "        # Check for 'prepare_model' \n",
    "        if \"prepare_model\" not in self.config:\n",
    "            raise KeyError(\"Key 'prepare_model' is missing from config.yaml!\")\n",
    "\n",
    "        prepare_model = self.config.prepare_model  \n",
    "        params = self.params\n",
    "\n",
    "        training_data = Path(training.training_data)\n",
    "\n",
    "        create_directories([Path(training.root_dir)])\n",
    "\n",
    "        training_config = TrainingConfig(\n",
    "            root_dir=Path(training.root_dir),\n",
    "            trained_model_path=Path(training.trained_model_path),\n",
    "            updated_base_model_path=Path(prepare_model.base_model_path),  \n",
    "            training_data=training_data,\n",
    "            params_epochs=params.EPOCHS,\n",
    "            params_batch_size=params.BATCH_SIZE,\n",
    "            params_is_augmentation=params.AUGMENTATION,\n",
    "            params_image_size=params.IMAGE_SIZE,\n",
    "            log_dir=Path(training.log_dir),\n",
    "            params_learning_rate =params.LEARNING_RATE\n",
    "        )\n",
    "\n",
    "        return training_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import tensorflow as tf\n",
    "from pathlib import Path\n",
    "from cnnClassifier.entity.config_entity import TrainingConfig\n",
    "from tensorflow.keras.preprocessing.image import ImageDataGenerator\n",
    "from cnnClassifier.logger import logger \n",
    "from sklearn.utils.class_weight import compute_class_weight\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "class Training:\n",
    "    def __init__(self, config: TrainingConfig):\n",
    "        self.config = config\n",
    "\n",
    "    def get_base_model(self):\n",
    "        \n",
    "        \"\"\"Load and compile the base model for training\"\"\"\n",
    "        logger.info(\"Loading base model...\")\n",
    "        self.model = tf.keras.models.load_model(self.config.updated_base_model_path)\n",
    "\n",
    "        # Compile the model\n",
    "        self.model.compile(\n",
    "            optimizer=tf.keras.optimizers.Adam(learning_rate=self.config.params_learning_rate),\n",
    "            loss='binary_crossentropy',\n",
    "            metrics=['accuracy']\n",
    "        )\n",
    "\n",
    "        logger.info(\"Base model loaded and compiled successfully.\")\n",
    "\n",
    "\n",
    "    def train_valid_generator(self):\n",
    "        \"\"\"Prepare training and validation data generators\"\"\"\n",
    "\n",
    "        logger.info(\"Initializing training and validation data generators...\")\n",
    "\n",
    "        # Image size must match the model input\n",
    "        target_img_size = (256, 256)\n",
    "\n",
    "        # This automatically rescales and splits the dataset\n",
    "        datagenerator_kwargs = dict(\n",
    "            rescale=1./255,         # normalize pixel values\n",
    "            validation_split=0.20   # 20% for validation\n",
    "        )\n",
    "\n",
    "        dataflow_kwargs = dict(\n",
    "            target_size=target_img_size,\n",
    "            batch_size=self.config.params_batch_size,\n",
    "            class_mode=\"binary\",    # Binary classification\n",
    "            interpolation=\"bilinear\",\n",
    "            color_mode=\"rgb\",       # Convert grayscale to RGB\n",
    "        )\n",
    "\n",
    "        # Validation data generator (no augmentation)\n",
    "        valid_datagenerator = ImageDataGenerator(**datagenerator_kwargs)\n",
    "\n",
    "        combined_training_dir = os.path.join(\"artifacts\", \"data_ingestion\", \"Combined_Training_Data\")\n",
    "\n",
    "        self.valid_generator = valid_datagenerator.flow_from_directory(\n",
    "            directory=combined_training_dir,\n",
    "            subset=\"validation\",\n",
    "            shuffle=False,\n",
    "            **dataflow_kwargs\n",
    "        )\n",
    "\n",
    "        # Training data generator (with optional augmentation)\n",
    "        if self.config.params_is_augmentation:\n",
    "            train_datagenerator = ImageDataGenerator(\n",
    "                rotation_range=20,\n",
    "                zoom_range=0.2,\n",
    "                width_shift_range=0.1,\n",
    "                height_shift_range=0.1,\n",
    "                shear_range=0.1,\n",
    "                horizontal_flip=True,\n",
    "                **datagenerator_kwargs\n",
    "            )\n",
    "        else:\n",
    "            train_datagenerator = ImageDataGenerator(**datagenerator_kwargs)\n",
    "\n",
    "        self.train_generator = train_datagenerator.flow_from_directory(\n",
    "            directory=combined_training_dir,\n",
    "            subset=\"training\",\n",
    "            shuffle=True,\n",
    "            **dataflow_kwargs\n",
    "        )\n",
    "\n",
    "        print(\"Training data classes:\", self.train_generator.class_indices)\n",
    "        print(\"Number of training samples:\", self.train_generator.samples)\n",
    "        print(\"Number of validation samples:\", self.valid_generator.samples)\n",
    "\n",
    "        logger.info(\"Data generators initialized successfully.\")\n",
    "\n",
    "\n",
    "\n",
    "    def train(self):\n",
    "        \"\"\"Train the model with logging, Early Stopping, and Class Weights\"\"\"\n",
    "        logger.info(\"Starting model training...\")\n",
    "\n",
    "        self.steps_per_epoch = self.train_generator.samples // self.train_generator.batch_size\n",
    "        self.validation_steps = self.valid_generator.samples // self.valid_generator.batch_size\n",
    "\n",
    "        # Ensure log directory exists\n",
    "        log_dir = Path(self.config.root_dir) / \"logs\"\n",
    "        log_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "        combined_training_dir = os.path.join(\"artifacts\", \"data_ingestion\", \"Combined_Training_Data\")\n",
    "\n",
    "        stone_dir = os.path.join(combined_training_dir, \"Stone\")\n",
    "        non_stone_dir = os.path.join(combined_training_dir, \"Non-Stone\")\n",
    "\n",
    "        if not os.path.exists(stone_dir) or not os.path.exists(non_stone_dir):\n",
    "            raise FileNotFoundError(\"One of the class folders (Stone or Non-Stone) is missing in Combined_Training_Data\")\n",
    "\n",
    "        stone_count = len(os.listdir(stone_dir))\n",
    "        non_stone_count = len(os.listdir(non_stone_dir))\n",
    "\n",
    "\n",
    "        # Compute class weights\n",
    "        class_weights = compute_class_weight(\n",
    "            class_weight=\"balanced\",\n",
    "            classes=np.array([0, 1]),\n",
    "            y=[0] * non_stone_count + [1] * stone_count\n",
    "        )\n",
    "\n",
    "        class_weight_dict = {0: class_weights[0], 1: class_weights[1]}\n",
    "\n",
    "        # Early Stopping Callback\n",
    "        early_stopping = tf.keras.callbacks.EarlyStopping(\n",
    "            monitor=\"val_loss\",\n",
    "            patience=5,\n",
    "            restore_best_weights=True\n",
    "        )\n",
    "\n",
    "        # TensorBoard Callback\n",
    "        tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=str(log_dir))\n",
    "\n",
    "        # Model Checkpoint Callback to save the best model during training\n",
    "        checkpoint = tf.keras.callbacks.ModelCheckpoint(\n",
    "            filepath=self.config.trained_model_path,\n",
    "            monitor=\"val_accuracy\",\n",
    "            save_best_only=True,\n",
    "            mode=\"max\",\n",
    "            verbose=1\n",
    "        )\n",
    "\n",
    "        try:\n",
    "            history = self.model.fit(\n",
    "            self.train_generator,\n",
    "            epochs=self.config.params_epochs,\n",
    "            steps_per_epoch=self.steps_per_epoch,\n",
    "            validation_steps=self.validation_steps,\n",
    "            validation_data=self.valid_generator,\n",
    "            class_weight=class_weight_dict,\n",
    "            callbacks=[early_stopping, tensorboard_callback, checkpoint]\n",
    "        )\n",
    "            \n",
    "            \n",
    "            logger.info(\"Model training completed successfully.\")\n",
    "\n",
    "            self.save_model(\n",
    "                path=self.config.trained_model_path,\n",
    "                model=self.model\n",
    "            )\n",
    "            logger.info(f\"Model saved at {self.config.trained_model_path}\")\n",
    "\n",
    "        except Exception as e:\n",
    "            logger.error(f\"Training failed due to: {e}\")\n",
    "            raise e\n",
    "        \n",
    "        # Plot training & validation accuracy/loss\n",
    "        try:\n",
    "            metrics_dir = Path(self.config.root_dir) / \"metrics\"\n",
    "            metrics_dir.mkdir(parents=True, exist_ok=True)\n",
    "            # Plot Accuracy\n",
    "            plt.figure(figsize=(8, 6))\n",
    "            plt.plot(history.history['accuracy'], label='Train Accuracy')\n",
    "            plt.plot(history.history['val_accuracy'], label='Validation Accuracy')\n",
    "            plt.title('Model Accuracy')\n",
    "            plt.xlabel('Epochs')\n",
    "            plt.ylabel('Accuracy')\n",
    "            plt.legend()\n",
    "            plt.grid(True)\n",
    "            plt.savefig(metrics_dir / 'accuracy.png')\n",
    "            plt.close()\n",
    "\n",
    "            # Plot Loss\n",
    "            plt.figure(figsize=(8, 6))\n",
    "            plt.plot(history.history['loss'], label='Train Loss')\n",
    "            plt.plot(history.history['val_loss'], label='Validation Loss')\n",
    "            plt.title('Model Loss')\n",
    "            plt.xlabel('Epochs')\n",
    "            plt.ylabel('Loss')\n",
    "            plt.legend()\n",
    "            plt.grid(True)\n",
    "            plt.savefig(metrics_dir / 'loss.png')\n",
    "            plt.close()\n",
    "\n",
    "            logger.info(f\"Training metrics plotted and saved to {metrics_dir}\")\n",
    "\n",
    "        except Exception as plot_err:\n",
    "            logger.warning(f\"Could not plot training metrics due to: {plot_err}\")\n",
    "\n",
    "        \n",
    "    def save_model(self, path: Path, model: tf.keras.Model):\n",
    "        \"\"\"Save the trained model to a given path.\"\"\"\n",
    "        model.save(path)\n",
    "        logger.info(f\"Model saved successfully at: {path}\")       \n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-04-05 13:11:05,233: INFO: common: yaml file: config\\config.yaml loaded successfully]\n",
      "[2025-04-05 13:11:05,236: INFO: common: yaml file: params.yaml loaded successfully]\n",
      "[2025-04-05 13:11:05,239: INFO: common: created directory at: artifacts]\n",
      "[2025-04-05 13:11:05,239: INFO: common: created directory at: artifacts\\training]\n",
      "[2025-04-05 13:11:05,241: INFO: 3633928156: Loading base model...]\n",
      "[2025-04-05 13:11:05,738: WARNING: hdf5_format: No training configuration found in the save file, so the model was *not* compiled. Compile it manually.]\n",
      "[2025-04-05 13:11:05,749: INFO: 3633928156: Base model loaded and compiled successfully.]\n",
      "[2025-04-05 13:11:05,749: INFO: 3633928156: Initializing training and validation data generators...]\n",
      "Found 7763 images belonging to 2 classes.\n",
      "Found 31058 images belonging to 2 classes.\n",
      "Training data classes: {'Non-Stone': 0, 'Stone': 1}\n",
      "Number of training samples: 31058\n",
      "Number of validation samples: 7763\n",
      "[2025-04-05 13:11:06,839: INFO: 3633928156: Data generators initialized successfully.]\n",
      "[2025-04-05 13:11:06,839: INFO: 3633928156: Starting model training...]\n",
      "Epoch 1/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.7068 - accuracy: 0.5723\n",
      "Epoch 1: val_accuracy improved from -inf to 0.50529, saving model to artifacts\\training\\model.h5\n",
      "970/970 [==============================] - 1531s 2s/step - loss: 0.7068 - accuracy: 0.5723 - val_loss: 0.7188 - val_accuracy: 0.5053\n",
      "Epoch 2/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.6448 - accuracy: 0.6272\n",
      "Epoch 2: val_accuracy improved from 0.50529 to 0.53809, saving model to artifacts\\training\\model.h5\n",
      "970/970 [==============================] - 1524s 2s/step - loss: 0.6448 - accuracy: 0.6272 - val_loss: 0.6919 - val_accuracy: 0.5381\n",
      "Epoch 3/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.6018 - accuracy: 0.6715\n",
      "Epoch 3: val_accuracy improved from 0.53809 to 0.63159, saving model to artifacts\\training\\model.h5\n",
      "970/970 [==============================] - 1528s 2s/step - loss: 0.6018 - accuracy: 0.6715 - val_loss: 0.6317 - val_accuracy: 0.6316\n",
      "Epoch 4/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.5423 - accuracy: 0.7248\n",
      "Epoch 4: val_accuracy improved from 0.63159 to 0.67317, saving model to artifacts\\training\\model.h5\n",
      "970/970 [==============================] - 1526s 2s/step - loss: 0.5423 - accuracy: 0.7248 - val_loss: 0.6226 - val_accuracy: 0.6732\n",
      "Epoch 5/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.4611 - accuracy: 0.7791\n",
      "Epoch 5: val_accuracy improved from 0.67317 to 0.81573, saving model to artifacts\\training\\model.h5\n",
      "970/970 [==============================] - 1507s 2s/step - loss: 0.4611 - accuracy: 0.7791 - val_loss: 0.4713 - val_accuracy: 0.8157\n",
      "Epoch 6/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.3765 - accuracy: 0.8283\n",
      "Epoch 6: val_accuracy did not improve from 0.81573\n",
      "970/970 [==============================] - 1495s 2s/step - loss: 0.3765 - accuracy: 0.8283 - val_loss: 0.4545 - val_accuracy: 0.7886\n",
      "Epoch 7/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.3180 - accuracy: 0.8639\n",
      "Epoch 7: val_accuracy improved from 0.81573 to 0.86751, saving model to artifacts\\training\\model.h5\n",
      "970/970 [==============================] - 1490s 2s/step - loss: 0.3180 - accuracy: 0.8639 - val_loss: 0.3618 - val_accuracy: 0.8675\n",
      "Epoch 8/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.2791 - accuracy: 0.8833\n",
      "Epoch 8: val_accuracy did not improve from 0.86751\n",
      "970/970 [==============================] - 1485s 2s/step - loss: 0.2791 - accuracy: 0.8833 - val_loss: 0.3822 - val_accuracy: 0.8009\n",
      "Epoch 9/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.2490 - accuracy: 0.8981\n",
      "Epoch 9: val_accuracy did not improve from 0.86751\n",
      "970/970 [==============================] - 1483s 2s/step - loss: 0.2490 - accuracy: 0.8981 - val_loss: 0.4476 - val_accuracy: 0.7708\n",
      "Epoch 10/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.2288 - accuracy: 0.9114\n",
      "Epoch 10: val_accuracy did not improve from 0.86751\n",
      "970/970 [==============================] - 1944s 2s/step - loss: 0.2288 - accuracy: 0.9114 - val_loss: 0.4917 - val_accuracy: 0.7492\n",
      "Epoch 11/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.2115 - accuracy: 0.9169\n",
      "Epoch 11: val_accuracy did not improve from 0.86751\n",
      "970/970 [==============================] - 1486s 2s/step - loss: 0.2115 - accuracy: 0.9169 - val_loss: 0.3976 - val_accuracy: 0.8107\n",
      "Epoch 12/40\n",
      "970/970 [==============================] - ETA: 0s - loss: 0.1996 - accuracy: 0.9230\n",
      "Epoch 12: val_accuracy did not improve from 0.86751\n",
      "970/970 [==============================] - 1477s 2s/step - loss: 0.1996 - accuracy: 0.9230 - val_loss: 0.4506 - val_accuracy: 0.8190\n",
      "[2025-04-05 18:19:02,075: INFO: 3633928156: Model training completed successfully.]\n",
      "[2025-04-05 18:19:02,402: INFO: 3633928156: Model saved successfully at: artifacts\\training\\model.h5]\n",
      "[2025-04-05 18:19:02,402: INFO: 3633928156: Model saved at artifacts\\training\\model.h5]\n",
      "[2025-04-05 18:19:02,788: INFO: 3633928156: Training metrics plotted and saved to artifacts\\training\\metrics]\n"
     ]
    }
   ],
   "source": [
    "try: \n",
    "    config = ConfigurationManager()\n",
    "    training_config = config.get_training_config()  \n",
    "    training = Training(config=training_config)\n",
    "    training.get_base_model()\n",
    "    training.train_valid_generator()\n",
    "    training.train()\n",
    "except Exception as e:\n",
    "    logger.exception(e)\n",
    "    raise e"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
